core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorPageSourceProvider.java

```diff
@@ -28,4 +30,31 @@ ConnectorPageSource createPageSource(
             ConnectorTableHandle table,
             List<ColumnHandle> columns,
             DynamicFilter dynamicFilter);
+
+    /**
+     * Returns unenforced (effective) predicate that {@link ConnectorPageSource} would use to filter split data.
+     * If split is completely filtered out, then this method should return {@link TupleDomain#none}.
+     */
+    default TupleDomain<ColumnHandle> getUnenforcedPredicate(
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            TupleDomain<ColumnHandle> dynamicFilter)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Prunes columns from predicate that are not effective in filtering split data.
+     * If split is completely filtered out by given predicate, then this
+     * method must return {@link TupleDomain#none}.
```

Comment:
This is a weird mix of column pruning and constraint pruning (i..e, a column-wise intersection and a all-or-none intersection of the constraint). It would be more general and easier to reason about if it were a pure intersection of the given tupledomain with the constraint guaranteed by the split.

I can't tell yet what's the right abstraction as there are no uses of it up to this commit -- I will revisit once I review the rest of the PR.

Reply:
> It would be more general and easier to reason about if it were a pure intersection of the given tupledomain with the constraint guaranteed by the split.

I mostly agree. However, in case of bucketing you cannot generate `TupleDomain` for bucketing column, yet you are still able to perform filtering.
The same applies to Iceberg transformation partitioning.

Generally, determining if contains relationship is simpler computationally than actually getting all intersecting values. In this case we actually don't need to know actual intersection, hence this method definition.